// grimoirejs-fundamental/lib/Components/MeshRendererComponent.js
import Component from "grimoirejs/lib/Node/Component";
export default class MeshRenderer extends Component {
    $awake() {
        this.getAttribute("targetBuffer").boundTo("_targetBuffer");
        this.getAttribute("layer").boundTo("_layer");
        this.getAttribute("drawOffset").boundTo("_drawOffset");
        this.getAttribute("drawCount").boundTo("_drawCount");
        this.getAttribute("geometry").boundTo("_geometry");
    }
    $mount() {
        this._transformComponent = this.node.getComponent("Transform");
        this._materialContainer = this.node.getComponent("MaterialContainer");
    }
    $render(args) {
        if (this._layer !== args.layer) {
            return;
        }
        if (!this._geometry || (!args.material && !this._materialContainer.ready)) {
            return; // material is not instanciated yet.
        }
        const renderArgs = {
            targetBuffer: this._targetBuffer,
            geometry: this._geometry,
            attributeValues: null,
            camera: args.camera.camera,
            transform: this._transformComponent,
            buffers: args.buffers,
            viewport: args.viewport,
            drawCount: this._drawCount,
            drawOffset: this._drawOffset
        };
        if (args.material) {
            renderArgs.attributeValues = args.materialArgs;
            args.material.draw(renderArgs);
        }
        else {
            renderArgs.attributeValues = this._materialContainer.materialArgs;
            this._materialContainer.material.draw(renderArgs);
        }
        this.companion.get("gl").flush();
    }
}
MeshRenderer.attributes = {
    geometry: {
        converter: "Geometry",
        defaultValue: "quad"
    },
    targetBuffer: {
        converter: "String",
        defaultValue: "default"
    },
    layer: {
        converter: "String",
        defaultValue: "default"
    },
    drawCount: {
        converter: "Number",
        defaultValue: Number.MAX_VALUE
    },
    drawOffset: {
        converter: "Number",
        defaultValue: 0
    }
};
