// grimoirejs-fundamental/lib/Components/TextureBufferComponent.js
import Texture2D from "../Resource/Texture2D";
import Component from "grimoirejs/lib/Node/Component";
export default class TextureBufferComponent extends Component {
    $mount() {
        this.buffer = new Texture2D(this.companion.get("gl"));
    }
    $unmount() {
        this.buffer.destroy();
        this.buffer = null;
    }
    $resizeBuffer(arg) {
        if (!this.getValue("name")) {
            throw new Error(`Attribute 'name' must be specified.`);
        }
        this.buffer.update(0, arg.width, arg.height, 0, WebGLRenderingContext.RGBA, WebGLRenderingContext.UNSIGNED_BYTE, null);
        arg.buffers[this.getValue("name")] = this.buffer;
    }
}
TextureBufferComponent.attributes = {
    name: {
        converter: "String",
        defaultValue: undefined
    },
    format: {
        converter: "Enum",
        defaultValue: WebGLRenderingContext.RGBA,
        table: {
            RGBA: WebGLRenderingContext.RGBA,
            RGB: WebGLRenderingContext.RGB
        }
    }
};
